### PR TITLE
Add async helpers, CombineLatest enumerable

### PR DESCRIPTION
**Pull request by ChrisPulman**
These changes extend the async API surface, document usage, and include tests to validate expected behavior and lifetime handling.

**What kind of change does this PR introduce?**

update

**What is the current behavior?**

Async features are not sufficiently extensive in some situations.

**What is the new behavior?**

Introduce a set of async-native helper operators that provide parity with the synchronous ReactiveExtensions surface and add an enumerable overload for CombineLatest. Files added:

- src/ReactiveUI.Extensions/Async/Operators/ParityHelpers.cs: Implements many helper operators (AsSignal, CatchIgnore / CatchAndReturn, DoOnSubscribe, DropIfBusy, LatestOrDefault, LogErrors, ObserveOnSafe/If, Pairwise, Partition, ReplayLastOnSubscribe, ThrottleDistinct, ScanWithInitial, DebounceUntil, GetMax/GetMin, CombineLatestValuesAreAllTrue/False, ForEach, Not, SkipWhileNull, Start, WhereTrue/False/IsNotNull and others).
- src/ReactiveUI.Extensions/Async/Operators/CombineLatestEnumerable.cs: Adds CombineLatest(IEnumerable<>) overloads with correct subscription and snapshot semantics.
- src/ReactiveUI.Extensions.Tests/Async/ParityOperatorTests.cs: Unit tests exercising key parity helpers (WhereIsNotNull, CombineLatestValuesAreAllTrue, GetMax, ScanWithInitial, Pairwise, Partition, DoOnSubscribe, CatchIgnore, Start).
- README.md: Documented the new helpers/operators and added usage examples.

**What might this PR break?**

**Please check if the PR fulfills these requirements**
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Other information**:

### CODE REVIEW

File: src/ReactiveUI.Extensions/Async/Operators/CombineLatestEnumerable.cs [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+                        return;
+                    }
+
+                    var currentIndex = index;
+                    _subscriptions[index] = await _sources[index].SubscribeAsync(
+                        (value, token) => OnNextAsync(currentIndex, value, token),
+                        OnErrorResumeAsync,
+                        result => OnCompletedAsync(currentIndex, result),
+                        linkedCts.Token);
+                }
```
[the remark this answers is not in the document]
Comment (ChrisPulman) [reply src/ReactiveUI.Extensions/Async/Operators/CombineLatestEnumerable.cs]: @copilot apply changes based on [this feedback](https://github.com/reactiveui/Extensions/pull/135#discussion_r3005503168)

**Review by glennawatson — DISMISSED** [review]
## Review: Allocations, Performance, API Design & Modern C#

Overall this is a solid set of parity helpers. The main concerns are around hot-path allocations in `CombineLatestEnumerable`, API surface choices (`IList<T>` vs `IReadOnlyList<T>`), and a few minor items. See inline comments for details.

---

### HIGH — Hot-path LinkedCTS allocation in CombineLatestEnumerable (lines 116, 136)

`CancellationTokenSource.CreateLinkedTokenSource` is called on **every single `OnNextAsync` and `OnErrorResumeAsync`**. Each linked CTS allocates, registers with both parent tokens, and must be disposed. For high-throughput streams this is significant GC pressure.

The linked CTS links the upstream `cancellationToken` with `_disposeCts.Token`, but `_disposed == 1` is already checked inside the lock on the very next line. Once `CompleteAsync` fires, `_disposeCts.Cancel()` will cancel any in-flight work anyway.

**Recommendation:** Create the linked CTS **once** in `SubscribeAsync` and reuse it for the subscription lifetime, or eliminate it entirely and rely on the existing `_disposed` guard + `_disposeCts.Cancel()` in `CompleteAsync`.

---

### MEDIUM — `IList<T>` return type implies mutability (CombineLatestEnumerable.cs line 25)

`TryCreateSnapshot` always returns a `T[]`, but the public API exposes `IObservableAsync<IList<T>>`. `IList<T>` implies consumers can call `.Add()`, `.Remove()`, etc. — which would throw `NotSupportedException` at runtime on the underlying array.

**Recommendation:** Use `IReadOnlyList<T>` instead — it's more honest about the immutable snapshot semantics, prevents accidental mutation, and is the idiomatic modern C# choice for indexed read-only collections.

---

### MEDIUM — Locking on `_disposeCts` (CombineLatestEnumerable.cs line 156)

`lock (_disposeCts)` works but is a code smell — locking on an object whose primary purpose isn't synchronization can be surprising and fragile. With `LangVersion` set to `preview` and targeting net9.0/net10.0, consider a dedicated `System.Threading.Lock` field:
```csharp
private readonly Lock _syncLock = new();
```
This gives the benefit of `Lock.EnterScope()` returning a ref struct, avoiding the monitor overhead.

---

### MEDIUM — `Partition` evaluates predicate twice per element (ParityHelpers.cs line 307)

```csharp
var shared = source.Publish().RefCount();
return (shared.Where(predicate), shared.Where(value => !predicate(value)));
```

Every element runs `predicate` twice. If the predicate is expensive, this doubles the cost. A single-pass implementation using `GroupBy(predicate)` or a custom operator would be more efficient. At minimum, document that the predicate must be pure and is invoked twice per element.

---

### LOW — Unnecessary `List<T>` allocation in GetMin/GetMax (ParityHelpers.cs lines 402, 419)

```csharp
List<IObservableAsync<T>> allSources = [source, .. sources];
```

Since `sources` is a `params` array and the final size is known (`1 + sources.Length`), this allocates a `List<T>` + copies. A direct array allocation would avoid the list overhead:
```csharp
var allSources = new IObservableAsync<T>[sources.Length + 1];
allSources[0] = source;
sources.CopyTo(allSources, 1);
```

---

### LOW — `LatestOrDefault` silently swallows first value if equal to default (ParityHelpers.cs line 170)

`source.StartWith(defaultValue).DistinctUntilChanged()` — if the source's first actual value equals `defaultValue`, `DistinctUntilChanged` will suppress it. This may surprise consumers counting emissions or using `Take(2)`. Worth documenting in the XML doc.

---

### LOW — Test file mixes concerns (ParityAndInfrastructureCoverageTests.cs)

This file tests parity operators but also includes tests for `Optional`, `AsyncGate`, `Heartbeat`, `Stale`, and Subject factory validation (lines 373–484). Consider splitting infrastructure tests into their own class.

Also notably missing: `CombineLatestEnumerable` edge-case tests — error propagation mid-stream, partial completion, and disposal during subscription.

File: src/ReactiveUI.Extensions/Async/Operators/CombineLatestEnumerable.cs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+                        result => OnCompletedAsync(currentIndex, result),
+                        cancellationToken);
+                }
+            }
+
+            public ValueTask DisposeAsync() => CompleteAsync(null);
+
+            private async ValueTask OnNextAsync(int index, T value, CancellationToken cancellationToken)
+            {
+                using var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _disposeCts.Token);
```
Comment (glennawatson) [inline src/ReactiveUI.Extensions/Async/Operators/CombineLatestEnumerable.cs]: **HIGH — Hot-path allocation: `CancellationTokenSource.CreateLinkedTokenSource` per `OnNextAsync`**

This creates and disposes a linked CTS on **every single value emission**. Each linked CTS allocates a CTS object, registers with both parent tokens, and must be disposed. For high-throughput streams this is significant GC pressure.

The linked CTS links the upstream `cancellationToken` with `_disposeCts.Token`, but `_disposed == 1` is already checked inside the lock on the very next line. Once `CompleteAsync` fires, `_disposeCts.Cancel()` will cancel any in-flight work anyway.

**Recommendation:** Create the linked CTS **once** in `SubscribeAsync` and reuse it for the subscription lifetime, or eliminate it entirely and rely on the existing `_disposed` guard + `_disposeCts.Cancel()` in `CompleteAsync`.

File: src/ReactiveUI.Extensions/Async/Operators/CombineLatestEnumerable.cs [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+                        return;
+                    }
+
+                    await _observer.OnNextAsync(snapshot, linkedCts.Token);
+                }
+            }
+
+            private async ValueTask OnErrorResumeAsync(Exception error, CancellationToken cancellationToken)
+            {
+                using var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _disposeCts.Token);
```
Comment (glennawatson) [inline src/ReactiveUI.Extensions/Async/Operators/CombineLatestEnumerable.cs]: Same hot-path linked CTS allocation concern as `OnNextAsync` above — this fires on every error resume call.

File: src/ReactiveUI.Extensions/Async/Operators/CombineLatestEnumerable.cs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+{
+    /// <summary>
+    /// Combines the latest value from each asynchronous observable sequence in the supplied collection.
+    /// </summary>
+    /// <typeparam name="T">The element type produced by the source sequences.</typeparam>
+    /// <param name="sources">The source sequences to combine.</param>
+    /// <returns>An observable sequence that emits a snapshot of the latest values whenever any source produces a new value,
+    /// after all sources have produced at least one value.</returns>
+    /// <exception cref="ArgumentNullException">Thrown if <paramref name="sources"/> is <see langword="null"/>.</exception>
+    public static IObservableAsync<IList<T>> CombineLatest<T>(this IEnumerable<IObservableAsync<T>> sources)
```
Comment (glennawatson) [inline src/ReactiveUI.Extensions/Async/Operators/CombineLatestEnumerable.cs]: **MEDIUM — `IList<T>` return type implies mutability**

`TryCreateSnapshot` always returns a `T[]`, but the public API exposes `IObservableAsync<IList<T>>`. `IList<T>` implies consumers can call `.Add()`, `.Remove()`, etc. — which would throw `NotSupportedException` at runtime on the underlying array.

**Recommendation:** Use `IReadOnlyList<T>` instead — it's more honest about the immutable snapshot semantics, prevents accidental mutation, and is the idiomatic modern C# choice for indexed read-only collections.

File: src/ReactiveUI.Extensions/Async/Operators/CombineLatestEnumerable.cs [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+            private ValueTask OnCompletedAsync(int index, ReactiveUI.Extensions.Async.Internals.Result result)
+            {
+                if (result.IsFailure)
+                {
+                    return CompleteAsync(result);
+                }
+
+                var shouldComplete = false;
+                lock (_disposeCts)
```
Comment (glennawatson) [inline src/ReactiveUI.Extensions/Async/Operators/CombineLatestEnumerable.cs]: **MEDIUM — Locking on `_disposeCts`**

`lock (_disposeCts)` works but is a code smell — locking on an object whose primary purpose isn't synchronization can be surprising and fragile. If CTS internals ever take their own lock, this risks deadlocks.

With `LangVersion` set to `preview` and targeting net9.0/net10.0, consider a dedicated `System.Threading.Lock` field:
```csharp
private readonly Lock _syncLock = new();
```
This gives the benefit of `Lock.EnterScope()` returning a ref struct, avoiding the monitor overhead.

**Review by glennawatson — CHANGES_REQUESTED** [review]
## Review: Allocations, Performance, API Design & Modern C#

Overall a solid set of parity helpers. See inline comments for specific concerns, ordered by priority.

**Review by glennawatson — COMMENTED** [review]
## Follow-up: LINQ on hot paths

A few System.Linq calls run inside CombineLatest selectors, meaning they execute on **every emission**. These allocate an enumerator per call through the `IList<T>` interface. Simple `for` loops would be zero-alloc.

File: src/ReactiveUI.Extensions/Async/Operators/CombineLatestEnumerable.cs (hunk 2 above)
Comment (ChrisPulman) [reply src/ReactiveUI.Extensions/Async/Operators/CombineLatestEnumerable.cs]: @copilot+claude-sonnet-4.6 create a Pull Request and analyse and resolve these concerns to remove the hot paths and provide a highly performant resolution.

**Review by glennawatson — APPROVED** [review]
